File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (root == NULL) return result;
        queue<TreeNode*> q;
        q.push(root);
        bool leftToRight = true;
        while (!q.empty()) {
            int size = q.size();
            vector<int> level(size);  // Vector to store current level in correct order

            for (int i = 0; i < size; i++) {
                TreeNode* currentNode = q.front();
                q.pop();
                int index = leftToRight ? i : size - i - 1;
                level[index] = currentNode->val;

                if (currentNode->left) {
                    q.push(currentNode->left);
                }
                if (currentNode->right) {
                    q.push(currentNode->right);
                }
            }
            // Add the current level to the result and flip the traversal direction
            result.push_back(level);
            leftToRight = !leftToRight;
        }
        return result;
        }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>(); // To store the final zigzag level order traversal
        if (root == null) {
            return result; // If the tree is empty, return an empty list
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root); // Start with the root
        boolean leftToRight = true; // Flag to control the zigzag pattern
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll(); // Dequeue the node
                
                if (leftToRight) {
                    level.add(node.val); // Left to right: Add normally
                } else {
                    level.add(0, node.val); // Right to left: Add to the beginning of the list
                }
                
                // Enqueue children
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            
            result.add(level); // Add the current level to the result
            leftToRight = !leftToRight; // Toggle the direction
        }
        
        return result; // Return the final zigzag level order traversal
    }
}